### PR TITLE
Add basic write support

### PR DESCRIPTION
**Pull request by pka**
This is a minimal writer implementation. Before implementing the missing parts, I would like to discuss the basic design.

Main questions:
1. Should we prepare for other backends (e.g. S3)?
   - (Trait could be done later)
2. Do we want an async API?
   - Probably depends on 1) 

Missing functionality:
* [x] Leaf directories
* [x] API for setting header data (center_zoom, etc.)
* [x] Configurable compression
* [x] Compression for tiles
* [x] Proper errors
* [x] Some TODOs and unwraps

### CODE REVIEW

**Comment by nyurik** [comment]
wrt 1 - you mean writing directly to s3?  Does s3 allow random access of the already open file (we would need to go back to the directory to update file positions i would think)

**Comment by pka** [comment]
AFAIK, S3 doesn't support seek operations or partial updates, but with a [multipart upload](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpu-upload-object.html), you can defer uploading header and root directory. So a direct upload to S3 would be possble. But this has no high priority for me.

**Comment by nyurik** [comment]
TIL, thx.  And lets follow [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it) - if someone wants it, they will add it :)

**Comment by pka** [comment]
This is now ready for a review.

I realized an important property, which I want to add as a followup: deduplicating non-subsequent tiles. This has performance drawbacks when reading, but probably reduces the file size substantially.

File: src/writer.rs [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+        let num_leaves = root_dir.entries().len();
+        Ok((root_dir, num_leaves))
+    }
+    fn dir_size(&self, entries: &[DirEntry]) -> PmtResult<usize> {
+        let dir = Directory::from_entries(entries);
+        dir.compressed_size(self.header.internal_compression)
+    }
+
+    /// Finish writing the `PMTiles` file.
+    pub fn finish(mut self) -> PmtResult<()> {
```
Comment (nyurik) [inline src/writer.rs:311]: If finalizing a file is required, should we add the `Drop` support, possibly panicing on error? Otherwise we could be left in a bad non-finalized state... unless of course user ends writing early due to some other error - in which case we are also left in a bad state, but we shouldn't finalize it...   Or perhaps we should delete the output?

We could also introduce `poison` concept - if any `write` method on this struct fail, we set a flag before returning an error, and in that case the file will not be finalized automatically. Otherwise, finalize the file on Drop unless it was already finalized by explicit call.

How does std::fs  handle errors during Drop?

File: src/writer.rs [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+        self.header.write_to(&mut self.out)?;
+        self.out.write_all(&root_dir_buf)?;
+        self.out.flush()?;
+
+        Ok(())
+    }
+}
+
+fn compare_blobs(a: &[u8], b: &[u8]) -> bool {
+    a.len() == b.len() && a.iter().zip(b.iter()).all(|(a, b)| a == b)
```
Comment (nyurik) [inline src/writer.rs:337]: why not `a == b` ?

File: src/writer.rs [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+}
+
+fn into_u32(v: usize) -> PmtResult<u32> {
```
Comment (nyurik) [inline src/writer.rs:340]: it seems this is only used twice - should it be a func?  I'm ok either way

**Review by nyurik — COMMENTED** [review]
looks pretty good!  I left a few thoughts

File: src/directory.rs [hunk 4]
```diff
@@ -17,6 +19,19 @@ impl Debug for Directory {
 }
 
 impl Directory {
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            entries: Vec::with_capacity(capacity),
+        }
+    }
+    pub(crate) fn from_entries(entries: &[DirEntry]) -> Self {
```
[the remark this answers is not in the document]
Comment (pka) [reply src/directory.rs:28]: There is no case for that yet. It's mainly used for temporary directories for optimizing leaves.

File: src/lib.rs [hunk 5]
```diff
@@ -1,3 +1,19 @@
+//! Read and write `PMTiles` according to the [PMTiles v3 spec](https://github.com/protomaps/PMTiles/blob/master/spec/v3/spec.md).
```
[the remark this answers is not in the document]
Comment (pka) [reply src/lib.rs:1]: I'm fine with both. Do the doc checks also run, when README.md is included?

File: src/writer.rs [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+            n_tile_entries: None,
+            n_tile_contents: None,
+            clustered: true,
+            internal_compression: Compression::Gzip,
+            tile_compression,
+            tile_type,
+            min_zoom: 0,
+            max_zoom: 22,
+            min_longitude: -180.0,
+            min_latitude: -85.0,
```
[the remark this answers is not in the document]
Comment (pka) [reply src/writer.rs]: Thanks for the hint. I took the numbers from the reference example, but your figure is more precise and the same as used by MapLibre JS.

File: src/writer.rs [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+        writer.header.metadata_length = metadata_length;
+        writer.header.data_offset = MAX_INITIAL_BYTES as u64 + metadata_length;
+
+        Ok(writer)
+    }
+}
+impl<W: Write + Seek> PmTilesStreamWriter<W> {
+    /// Add tile to writer
+    /// Tiles are deduplicated and written to output.
+    /// `tile_id` should be increasing.
```
[the remark this answers is not in the document]
Comment (pka) [reply src/writer.rs]: Adressed in 4586c4c

File: src/writer.rs (hunk 1 above)
Comment (pka) [reply src/writer.rs:339]: I'm not sure about that. I see the point, but don't think that we should put too much effort into fixing an obvious user bug.

File: src/writer.rs (hunk 2 above)
Comment (pka) [reply src/writer.rs]: You are kidding me???

File: src/writer.rs (hunk 3 above)
Comment (pka) [reply src/writer.rs:364]: Maybe it also useful for the reader, where currently `allow(clippy::cast_possible_truncation)` is set.

File: src/directory.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nyurik) [reply src/directory.rs:28]: let me rephrase - per Rust recommended usage, it is always better to "ask for ownership if you need to own it" - this way the caller will have the choice of cloning or passing the value's ownership if no longer needed. Here this is an internal API, so it is up to you.

File: src/writer.rs (hunk 1 above)
Comment (nyurik) [reply src/writer.rs:339]: ok, lets keep as is.  I tend to use the word `finalize`, but either works

**Comment by pka** [comment]
Regarding the failed semver check:

```
--- failure enum_variant_added: enum variant added on exhaustive enum ---

Description:
A publicly-visible enum without #[non_exhaustive] has a new variant.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#enum-variant-new
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/enum_variant_added.ron

Failed in:

     Summary semver requires new major version: 1 major and 0 minor checks failed
  variant PmtError:IndexEntryOverflow in /home/runner/work/pmtiles-rs/pmtiles-rs/src/error.rs:28
```

Shall I add `#[non_exhaustive]` or increase the version?

**Comment by nyurik** [comment]
Might as well bump the version

**Comment by nyurik** [comment]
Why do you want to make it non-exhaustive? I am ok to break compat - not a biggie tbh

**Comment by nyurik** [comment]
but come to think of it, errors are a pretty good candidate for non-exhaustive, so yeah, lets keep it going forward... but enabling it would still be a breaking change, right?

**Comment by nyurik** [comment]
one more thing - i think `write` support is a less used feature, and since it requires extra dependencies, lets make this into an optional feature

**Comment by pka** [comment]
> one more thing - i think `write` support is a less used feature, and since it requires extra dependencies, lets make this into an optional feature

`flate2` is already an indirect dependency and `countio` is quite small. So I thougt it's not worth the hassle.

**Comment by lseelenbinder** [comment]
Ohhh! This is great. Just seeing the PR now for some reason. I'll give it a review shortly. Thank you @pka!

File: README.md [hunk 8]
```diff
@@ -17,16 +17,31 @@ originally created by Brandon Liu for Protomaps.
...
-- [ ] Support writing and conversion to and from MBTiles + `x/y/z`
+- [ ] Support conversion to and from MBTiles + `x/y/z`
 - [ ] Support additional backends (sync `mmap` and `http` at least)
 - [ ] Support additional async styles (e.g., `async-std`)
 
 PRs welcome!
 
+## Usage examples
+
+### Writing a PMTiles file
```
Comment (nyurik) [inline README.md:33]: We must add a read example here before merging, or else it will get really confusing why we have a write without read

File: README.md [hunk 9]
```diff
@@ -17,16 +17,31 @@ originally created by Brandon Liu for Protomaps.
...
+
+```rust
+use pmtiles::{tile_id, PmTilesWriter, TileType};
+use std::fs::File;
+
+let file = File::create("tiles.pmtiles").unwrap();
+let mut writer = PmTilesWriter::new(TileType::Mvt).create(file).unwrap();
+writer.add_tile(tile_id(0, 0, 0), &[/*...*/]).unwrap();
```
Comment (nyurik) [inline README.md:41]: do we allow zero-length tiles?

File: src/lib.rs [hunk 10]
```diff
@@ -15,8 +17,8 @@ pub mod cache;
 mod directory;
 mod error;
 mod header;
-#[cfg(feature = "__async")]
 mod tile;
+mod writer;
```
Comment (nyurik) [inline src/lib.rs:21]: could you clarify why cfg(__async) was removed here?  We really ought to make this PR behind a feature flag

Comment (pka) [reply src/lib.rs:21]: I was glad to reduce the `cfg(__)` complexity a bit with 72f57629 - it reminded me of the bad old C/++ `#ifndef` days. IMO, a pmtile library should come with its core functionality (like calculating a tile id) and dependencies (like flate2) by default. I understand the feature flags for the different backend implementations with heavy or specialized dependencies, but not for everything. That said, if the maintainers prefer a feature flag for writing pmtile files, I can give it a try.

Comment (nyurik) [reply src/lib.rs:21]: I agree that we shouldn't go overboard on features.  My only concern is that if we have additional dependencies due to write support, it is better to make write - optional, simply because write is a relatively rare usage compare to read. My understanding is that you need compression, which might be a significant extra compilation burden - and that IMO should not be levied on users unless needed.

**Review by lseelenbinder — CHANGES_REQUESTED** [review]
First of all, thanks for setting this all up! I'm glad we're getting this functionality added.

I'd like to understand how much impact having the leaf directories at the end of the archive would have on performance. Hopefully it's not too much, because this elegantly solves the need for a temporary file.

Another option is to reserve N bytes at the beginning of every archive for the header + directories, but that's suboptimal as well.

**Comment by pka** [comment]
> I'd like to understand how much impact having the leaf directories at the end of the archive would have on performance. Hopefully it's not too much, because this elegantly solves the need for a temporary file.
> 
> Another option is to reserve N bytes at the beginning of every archive for the header + directories, but that's suboptimal as well.

I would not expect a performance impact, even via http. An optimization for initial loading would be adding the entries for z0-z2 to the root directory, which is a TODO copied from the pmtiles-go implementation. Shouldn't be difficult to add to the current implementation.

File: README.md (hunk 9 above)
Comment (pka) [reply README.md:41]: Zero-length tiles added with `add_tile` are ignored. The spec doesn't allow storing zero-length tiles and I think throwing an error is not absolutely necessary.

File: src/writer.rs [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+            metadata_length: 0,
+            leaf_offset: 0,
+            leaf_length: 0,
+            data_offset: 0,
+            data_length: 0,
+            n_addressed_tiles: None,
+            n_tile_entries: None,
+            n_tile_contents: None,
+            clustered: true,
+            internal_compression: Compression::Gzip,
```
[the remark this answers is not in the document]
Comment (pka) [reply src/writer.rs:94]: That's the default value which can be overridden with the `internal_compression` builder function. But currently, only compression methods supported by the reader are included.

File: src/writer.rs [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+        };
+        Self {
+            header,
+            metadata: "{}".to_string(),
+        }
+    }
+
+    /// Set the compression for metadata and directories.
+    #[must_use]
+    pub fn internal_compression(mut self, compression: Compression) -> Self {
```
[the remark this answers is not in the document]
Comment (pka) [reply src/writer.rs:115]: `PmTilesWriter` *is* a builder struct, generating a `PmTilesStreamWriter` struct with `create`. `Builder` is not in the name, because the resulting code looks better without it, IMO.

File: src/writer.rs [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+
+        Ok(writer)
+    }
+}
+
+impl<W: Write + Seek> PmTilesStreamWriter<W> {
+    /// Add tile to writer.
+    ///
+    /// Tiles are deduplicated and written to output.
+    /// `tile_id` should be increasing for best read performance.
```
[the remark this answers is not in the document]
Comment (pka) [reply src/writer.rs:215]: In contrary to the https://github.com/arma-place/pmtiles-rs writer, tiles are written directly to disk an not held in memory and sorted before writing. So `add_tile` expects the tiles already in correct order, which means `tile_id` has to be known anyway.

File: src/writer.rs [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+            v => Err(UnsupportedCompression(v))?,
+        }
+        Ok(())
+    }
+
+    fn write_compressed_to_counted<W: Write>(
+        &self,
+        writer: &mut Counter<W>,
+        compression: Compression,
+    ) -> PmtResult<usize> {
```
[the remark this answers is not in the document]
Comment (pka) [reply src/writer.rs:52]: In consequence, the reader(s) should then also have its own error enum(s).

File: src/writer.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (lseelenbinder) [reply src/writer.rs:215]: Ah, got it. I was mixing up `tile_id` with something else—forgetting it's just the (z,x,y) on a hilbert curve.

File: src/writer.rs [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+    pub fn create<W: Write + Seek>(self, writer: W) -> PmtResult<PmTilesStreamWriter<W>> {
+        let mut out = Counter::new(BufWriter::new(writer));
+
+        // We use the following layout:
+        // +--------+----------------+----------+-----------+------------------+
+        // |        |                |          |           |                  |
+        // | Header | Root Directory | Metadata | Tile Data | Leaf Directories |
+        // |        |                |          |           |                  |
+        // +--------+----------------+----------+-----------+------------------+
+        // This allows writing without temporary files. But it requires Seek support.
```
[the remark this answers is not in the document]
Comment (lseelenbinder) [reply src/writer.rs:185]: I see. I think you're right. The root directory is usually read as part of the header, and then the leaf directories are clustered, just as they would be normally.

This is a pretty clever solution, to be honest. I'm curious what @bdon thinks about this for go-pmtiles?

File: README.md (hunk 9 above)
Comment (lseelenbinder) [reply README.md:41]: I agree.

File: src/writer.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (lseelenbinder) [reply src/writer.rs:115]: If we switched to `pmtiles::writer::StreamWriter` then we could have `pmtiles::writer::Builder` without confusion.

File: src/writer.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (bdon) [reply src/writer.rs:185]: leaf directories at the end is in-spec and also how the java implementation in planetiler works: 

https://github.com/onthegomap/planetiler/blob/main/planetiler-core/src/main/java/com/onthegomap/planetiler/pmtiles/WriteablePmtiles.java#L220

**Comment by lseelenbinder** [comment]
> wrt 1 - you mean writing directly to s3? Does s3 allow random access of the already open file (we would need to go back to the directory to update file positions i would think)

I think this would be theoretically possible if we use multi-part upload, and refrain from uploading the first part until the end, but I'd have to check the docs to confirm this is possible.

File: src/writer.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (lseelenbinder) [reply src/writer.rs:115]: I _think_ we can accomplish the same thing with name spacing: `pmtiles::writer::StreamBuilder` and `pmtiles::writer::StreamWriter`. It's a pretty common pattern in Rust, and I think keeps things concise and manageable.

Any thoughts on this @nyurik?

File: src/writer.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (lseelenbinder) [reply src/writer.rs:52]: That isn't a bad thing, in my opinion. The Error type is quite packed right now.

I think `OpenError`, `ReadError`, and `WriteError` with associated `Result` types makes sense at this point, and leads to better code and a clearer API. @nyurik did a lot of the Error-related work, so his input is appreciated.

**Comment by lseelenbinder** [comment]
When I have the chance, I was planning to grab a copy of this PR and do some additional cleanup that makes sense now. I'll make it dependent against your code, @pka, so we don't have merge issues.

I'm struggling with how much to ask for in this PR vs what we should clean up after. Let's *not* immediately release after merging to give us some time to clean up the APIs in a more general way.

File: src/writer.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (nyurik) [reply src/writer.rs:115]: It took me a few seconds to understand "name spacing" meant "namespacing" :D  Naming is hard...

I am not a big fan of mega-generic names like `Builder` by itself, nor adding namespace as part of the required pattern... I do like `PmTilesWriter::new(...).create(...)` as it is cleaner to the user.  We could name the resulting struct as `PmTilesWriterStream` or `PmTilesStream` or whatever.

File: src/writer.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (nyurik) [reply src/writer.rs:94]: on naming - `internal` is somewhat confusing - is this the header and/or metadata compression?

File: src/writer.rs [hunk 16]
```diff
@@ -0,0 +1,492 @@
...
+        writer.write_all(self)
+    }
+}
+
+impl PmTilesWriter {
+    /// Create a new `PMTiles` writer with default values.
+    #[must_use]
+    pub fn new(tile_type: TileType) -> Self {
+        let tile_compression = match tile_type {
+            TileType::Mvt => Compression::Gzip,
```
[the remark this answers is not in the document]
Comment (nyurik) [reply src/writer.rs:76]: I do too, but let's limit the scope of this PR :)

File: src/writer.rs [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+    metadata: String,
+}
+
+/// `PMTiles` streaming writer.
+pub struct PmTilesStreamWriter<W: Write + Seek> {
+    out: Counter<BufWriter<W>>,
+    header: Header,
+    entries: Vec<DirEntry>,
+    n_addressed_tiles: u64,
+    // TODO: Replace with digest HashMap for deduplicating non-subsequent tiles
```
[the remark this answers is not in the document]
Comment (nyurik) [reply src/writer.rs:24]: I agree, but lets try to get this merged with MVP functionality first, or else we risk dragging this PR forever :)

File: README.md (hunk 8 above)
Comment (nyurik) [reply README.md:33]: Come to think of it - agree, this is not a blocker for merging

File: src/writer.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (pka) [reply src/writer.rs:94]: Thats the spec naming: https://github.com/protomaps/PMTiles/blob/main/spec/v3/spec.md#internal-compression-ic

File: src/writer.rs [hunk 18]
```diff
@@ -0,0 +1,492 @@
...
+                    dir.push(entry);
+                }
+                dir.write_compressed_to(&mut self.out, self.header.internal_compression)?;
+            }
+        }
+        Ok(root_dir)
+    }
+
+    fn optimize_directories(&self, target_root_len: usize) -> PmtResult<(Directory, usize)> {
+        // Same logic as go-pmtiles optimizeDirectories
```
[the remark this answers is not in the document]
Comment (pka) [reply src/writer.rs]: Links added.

File: src/writer.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (lseelenbinder) [reply src/writer.rs:24]: Agreed.

File: src/writer.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (lseelenbinder) [reply src/writer.rs:115]: Okay, this is a pretty fundamental decision—so I'd say this blocks the MR until we decide.

I did some research, and here's my suggestion.

The "thing" we want to build is called `PmTilesWriter` (I think "stream" is an implementation—99% of use-cases prefer streaming anyways.).

Then we can have a `Builder` in the same module, but hide the complexity behind a `PmTilesWriter::builder()` call.

This results in the following pseudocode:

```rust
let writer = PmTilesWriter::builder(tile_type).create("file.pmtiles");
```

The user never needs to know about the underlying builder structure, and it keeps the naming manageable. I'm happy to do the refactor, but just want to make sure there's no strong objections.

File: src/writer.rs [hunk 19]
```diff
@@ -0,0 +1,492 @@
...
+        Ok(writer)
+    }
+}
+
+impl<W: Write + Seek> PmTilesStreamWriter<W> {
+    /// Add tile to writer.
+    ///
+    /// Tiles are deduplicated and written to output.
+    /// `tile_id` should be increasing for best read performance.
+    pub fn add_tile(&mut self, tile_id: u64, data: &[u8]) -> PmtResult<()> {
```
[the remark this answers is not in the document]
Comment (lseelenbinder) [reply src/writer.rs:216]: Let's make this a TODO for after merging.

**Comment by lseelenbinder** [comment]
Okay! We're close. Just have to resolve the naming detail, and I'll be happy to do a cleanup pass after merging.

File: src/writer.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (nyurik) [reply src/writer.rs:115]: @lseelenbinder I just realized you were waiting for me here - this looks fine, want to make that change?  Lets merge and push this!

**Comment by nyurik** [comment]
ok, lots of changes here - write is now feature-gated, lots of other minor cleanups and changes.  This PR relies on #56, but once merged, should be easier to review.

File: src/writer.rs [hunk 20]
```diff
@@ -0,0 +1,492 @@
...
+impl PmTilesWriter {
+    /// Create a new `PMTiles` writer with default values.
+    #[must_use]
+    pub fn new(tile_type: TileType) -> Self {
+        let tile_compression = match tile_type {
+            TileType::Mvt => Compression::Gzip,
+            _ => Compression::None,
+        };
+        #[allow(clippy::excessive_precision, clippy::unreadable_literal)]
+        let header = Header {
```
[the remark this answers is not in the document]
Comment (nyurik) [reply src/writer.rs]: i made it into a `new` with a few params

**Review by nyurik — COMMENTED** [review]
@lseelenbinder i made a few more refactorings. Any blockers to merge this?

**Review by lseelenbinder — APPROVED** [review]
LGTM!

I think there's quite a few low-hanging improvements for future PRs, but this looks solid as a first release.
